File: core/RemoveNthNodeFromEndOfList/RemoveNthNodeFromEndOfList.java
package com.dbc;

public class RemoveNthNodeFromEndOfList {
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode preHead = new ListNode();
        preHead.next = head;
        ListNode res = preHead;
        while (n > 1) {
            head = head.next;
            n--;
        }
        while (head.next != null) {
            head = head.next;
            preHead = preHead.next;
        }
        preHead.next = preHead.next.next;
        return res.next;
    }
}

File: core/RemoveNthNodeFromEndOfList/RemoveNthNodeFromEndOfList.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def removeNthFromEnd(head: ListNode, n: int) -> ListNode:
    pre_node = ListNode()
    pre_node.next = head
    point = head
    res = pre_node
    while n > 1:
        point = point.next
        n -= 1

    while point.next:
        point = point.next
        pre_node = pre_node.next

    pre_node.next = pre_node.next.next

    return res.next


